collectors/ebpf.plugin/ebpf_hardirq.c

```diff
@@ -0,0 +1,503 @@
+static int hardirq_write_dims(void *entry, void *data)
+{
+    UNUSED(data);
+
+    hardirq_val_t *v = entry;
+
+    // IRQs get dynamically added in, so add the dimension if we haven't yet.
+    if (!v->dim_exists) {
+        ebpf_write_global_dimension(
+            v->name, v->name,
+            ebpf_algorithms[NETDATA_EBPF_INCREMENTAL_IDX]
+        );
+        v->dim_exists = true;
+    }
+
+    write_chart_dimension(v->name, v->latency);
+
+    return 1;
+}
+
+static void hardirq_write_static_dims()
+{
+    uint32_t i;
+    for (i = 0; i < HARDIRQ_EBPF_STATIC_END; i++) {
+        write_chart_dimension(
+            hardirq_static_vals[i].name,
+            hardirq_static_vals[i].latency
+        );
+    }
+}
+
+/**
+* Main loop for this collector.
+*/
+static void hardirq_collector(ebpf_module_t *em)
+{
+    hardirq_ebpf_vals = callocz(
+        (running_on_kernel < NETDATA_KERNEL_V4_15) ? 1 : ebpf_nprocs,
+        sizeof(hardirq_ebpf_val_t)
+    );
+    hardirq_ebpf_static_vals = callocz(
+        (running_on_kernel < NETDATA_KERNEL_V4_15) ? 1 : ebpf_nprocs,
+        sizeof(hardirq_ebpf_static_val_t)
+    );
+
+    avl_init_lock(&hardirq_pub, hardirq_val_cmp);
+
+    // create reader thread.
+    hardirq_threads.thread = mallocz(sizeof(netdata_thread_t));
+    hardirq_threads.start_routine = hardirq_reader;
+    netdata_thread_create(
+        hardirq_threads.thread,
+        hardirq_threads.name,
+        NETDATA_THREAD_OPTION_JOINABLE,
+        hardirq_reader,
+        em
+    );
+
+    // create chart and static dims.
+    pthread_mutex_lock(&lock);
+    hardirq_create_charts();
+    hardirq_create_static_dims();
+    pthread_mutex_unlock(&lock);
+
+    // loop and read from published data until ebpf plugin is closed.
+    while (!close_ebpf_plugin) {
+        pthread_mutex_lock(&collect_data_mutex);
+        pthread_cond_wait(&collect_data_cond_var, &collect_data_mutex);
+        pthread_mutex_lock(&lock);
+
+        // write dims now for all hitherto discovered IRQs.
+        write_begin_chart("system", "hardirq_latency");
+        avl_traverse_lock(&hardirq_pub, hardirq_write_dims, NULL);
+        hardirq_write_static_dims();
+        write_end_chart();
+
+        pthread_mutex_unlock(&lock);
+        pthread_mutex_unlock(&collect_data_mutex);
+    }
+}
+
+/*****************************************************************
+ *  EBPF HARDIRQ THREAD
+ *****************************************************************/
+
+/**
+ * Hard IRQ latency thread.
+ *
+ * @param ptr a `ebpf_module_t *`.
+ * @return always NULL.
+ */
+void *ebpf_hardirq_thread(void *ptr)
+{
+    netdata_thread_cleanup_push(hardirq_cleanup, ptr);
+
+    ebpf_module_t *em = (ebpf_module_t *)ptr;
+    em->maps = hardirq_maps;
+
+    fill_ebpf_data(&hardirq_data);
+
+    if (!em->enabled) {
+        goto endhardirq;
+    }
+
+    if (ebpf_update_kernel(&hardirq_data)) {
+        goto endhardirq;
+    }
+
+    bool one_tp_succeeded = false;
```

Comment:
Another suggestion here, the whole block to enable the IRQ can go inside a specific function.

Reply:
In this particular case I don't know if it's worth it - the function will never be re-used, and just causes the reader to have to jump somewhere else.

Reply:
Actually, I think I can make it reusable by making it operate on `ebpf_tracepoint_t[]` and return how many tracepoint enables succeeded.

Thanks for the idea!